### PR TITLE
Closes #2231  added check in `derive_vars_period()` to remove unused columns

### LINKED ISSUE #2231: Bug: `derive_vars_period()` errors unexpectedly when `dataset_ref` has unused variables
### What happened?

When `period_ref` contains variables that are then not mentioned within the `new_vars` call, `derive_vars_period()` errors with the unhelpful error message: 
```
Error in `signal_duplicate_records()`:
! Dataset `dataset_add` contains duplicate records with respect to `STUDYID` and `USUBJID`
Run `get_duplicates_dataset()` to access the duplicate records
```
when really one would at least expect the error message to flag why this is occurring. See reproducible example for details.

### Session Information

Not relevant

### Reproducible Example

adsl <- tibble(STUDYID = "xyz", USUBJID = c("1", "2"))

# Add period variables to ADSL
period_ref <- tribble(
  ~USUBJID, ~APERIOD, ~APERSDT,     ~APEREDT,     ~APERSDTM,                ~APEREDTM,
  "1",             1, "2021-01-04", "2021-02-06", "2021-01-04 10:10:00", "2021-02-07 11:20:01", 
  "1",             2, "2021-02-07", "2021-03-07", "2021-02-07 11:20:02", "2021-03-07 10:10:00",
  "2",             1, "2021-02-02", "2021-03-02", "2021-02-02 12:20:01", "2021-03-03 13:10:01",
  "2",             2, "2021-03-03", "2021-04-01", "2021-03-03 13:10:02", "2021-04-01 10:10:00"
) %>%
  mutate(
    STUDYID = "xyz",
    APERIOD = as.integer(APERIOD),
    APERSDT = ymd(APERSDT),
    APERSDT = ymd(APEREDT),
    APERSDTM = ymd_hms(APERSDTM),
    APEREDTM = ymd_hms(APEREDTM),
  )

# This errors because we are not mentioning APERSDTM, APEREDTM anywhere
derive_vars_period(
  adsl,
  dataset_ref = period_ref,
  new_vars = exprs(APxxSDT = APERSDT, APxxEDT = APEREDT)
)

# This doesn't error because we mentioning APERSDTM, APEREDTM
derive_vars_period(
  adsl,
  dataset_ref = period_ref,
  new_vars = exprs(APxxSDT = APERSDT, APxxEDT = APEREDT, APxxSDTM = APERSDTM, APxxEDTM = APEREDTM)
)

# This doesn't error because we are removing APERSDTM, APEREDTM from period_ref
derive_vars_period(
  adsl,
  dataset_ref = period_ref %>% select(-ends_with("M")),
  new_vars = exprs(APxxSDT = APERSDT, APxxEDT = APEREDT)
)

### PR DESCRIPTION
**Pull request by manciniedoardo**
Closes #2231

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by manciniedoardo** [comment]
Issue was arising because `derive_vars_period()` does a lot of dynamic transposing/renaming with `dataset_ref` and when the latter contains variables not referenced anywhere in the function call, the transposing/renaming fails. Thus I added two lines of code to strip `dataset_ref` of any variables not relevant for the period derivation right at the start.

File: tests/testthat/test-period_dataset.R [hunk 1]
```diff
@@ -421,3 +421,44 @@ test_that("create_period_dataset Test 12: DT and DTM columns exist, pulls only o
...
+
+## Test 13: Periods derived even when some variables in dataset_ref are unused ----
+test_that("derive_vars_period Test 13: periods", {
+  expected <- tibble::tribble(
+    ~USUBJID, ~AP01SDT,     ~AP01EDT,
+    "1",      "2021-01-04", "2021-02-06",
+    "2",      "2021-02-02", "2021-03-02"
+  ) %>%
+    mutate(
+      dplyr::across(matches("AP\\d\\d[ES]DT"), ymd)
```
Comment (bms63) [inline tests/testthat/test-period_dataset.R:433]:
```suggestion
      across(matches("AP\\d\\d[ES]DT"), ymd)
```

Comment (bms63) [reply tests/testthat/test-period_dataset.R:433]: we already import dplry no need to `::` - we do this for tibble as it is a suggest package

File: R/period_dataset.R [hunk 2]
```diff
@@ -408,10 +408,15 @@ derive_vars_period <- function(dataset,
   } else {
     id_vars <- exprs(APHASEN)
   }
-  assert_data_frame(dataset_ref, required_vars = expr_c(subject_keys, new_vars, id_vars))
```
Comment (bundfussr) [inline R/period_dataset.R:411]: I would keep this (and remove `assert_data_frame()` calls in line 359 and 416). At the moment the argument name is not stated in the error message but it should (see https://github.com/pharmaverse/admiraldev/issues/352). Once https://github.com/pharmaverse/admiraldev/issues/352 is implemented the call in line 416 would refer to a dataset unknown by the users.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/period_dataset.R (hunk 2 above)
Comment (manciniedoardo) [reply R/period_dataset.R:411]: Actioned

**Review by bundfussr — APPROVED** [review]

**Review by bms63 — APPROVED** [review]

**Pull request merged**